features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/sha/sha1_alt.c

```diff
@@ -51,6 +51,21 @@ void mbedtls_sha1_free(mbedtls_sha1_context *ctx)
 void mbedtls_sha1_clone(mbedtls_sha1_context *dst,
                         const mbedtls_sha1_context *src)
 {
+    // If dst is H/W context, we need to change it to S/W context first before cloning to.
+    while (dst->ishw) {
+        mbedtls_sha1_free(dst);
+        // We just want S/W context, so we lock SHA H/W resource if it is available.
+        if (crypto_sha_acquire()) {
+            mbedtls_sha1_init(dst);
+            crypto_sha_release();
+        }
+        else {
+            mbedtls_sha1_init(dst);
```

Comment:
Thanks for reworking. Wouldn't it be easier here to directly set up `dst` via `mbedtls_sha1_sw_init`?

Reply:
@hanno-arm Yes. But if I set up `dst` via `mbedtls_sha1_sw_init`, I need to set `ishw` to 0 additionally. If new context management is introduced (in `mbedtls_sha1_init`/`mbedtls_sha1_free`) in the future, I need also to take care of new modification accordingly similar to `ishw`. So I set up `dst` via `mbedtls_sha1_init` rather than `mbedtls_sha1_sw_init` to simplify future modification with context management.

Reply:
@ccli8 Fair point. Then what about making the present `mbedtls_sha1_init` a stub calling another (static, perhaps inline) routine `sha1_init_internal( mbedtls_sha1_context *ctx, int try_hw )` which accepts an additional flag indicating whether acquiring the lock and setting up a H/W context should be attempted? This function could be called from `mbedtls_sha1_init` with argument `1` and with argument `0` from `mbedtls_sha1_clone` then.

Reply:
@hanno-arm Following your suggestion, I update it in https://github.com/ARMmbed/mbed-os/pull/4925/commits/d6b0f0afb0a0a9da7220ef3115179ae383f2dde8.